op-node/rollup/derive/batch_queue.go

```diff
@@ -122,196 +97,140 @@ func (bq *BatchQueue) AddBatch(batch *BatchData) {
-				bq.log.Warn("duplicate batch", "epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions))
-				return
-			}
-		}
-	} else {
-		bq.log.Debug("First seen batch", "epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions))
-	}
-	// May have duplicate block numbers or individual fields, but have limited complete duplicates
-	bq.batchesByTimestamp[batch.Timestamp] = append(batches, &data)
-}
-
-// validExtension determines if a batch follows the previous attributes
-func (bq *BatchQueue) validExtension(batch *BatchWithL1InclusionBlock, prevTime, prevEpoch uint64) (valid bool, err error) {
-	if batch.Batch.Timestamp != prevTime+bq.config.BlockTime {
-		bq.log.Debug("Batch does not extend the block time properly", "time", batch.Batch.Timestamp, "prev_time", prevTime)
-		return false, nil
-	}
-	if batch.Batch.EpochNum != rollup.Epoch(prevEpoch) && batch.Batch.EpochNum != rollup.Epoch(prevEpoch+1) {
-		bq.log.Debug("Batch does not extend the epoch properly", "epoch", batch.Batch.EpochNum, "prev_epoch", prevEpoch)
-		return false, nil
-	}
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
-	l1BlockRef, err := bq.dl.L1BlockRefByNumber(ctx, batch.Batch.Epoch().Number)
-	cancel()
-	if err != nil {
-		return false, err
-	}
-
-	if l1BlockRef.Hash != batch.Batch.EpochHash {
-		bq.log.Debug("Batch epoch hash does not match expected L1 block hash", "batch_epoch", batch.Batch.Epoch(), "expected", l1BlockRef.ID())
-		return false, nil
+	validity := CheckBatch(bq.config, bq.log, bq.l1Blocks, bq.next.SafeL2Head(), &data)
+	if validity == BatchDrop {
+		log.Warn("ingested invalid batch from L1, dropping it instead of buffering it")
+		return
 	}
-
-	// Note: `Batch.EpochNum` is an external input, but it is constrained to be a reasonable size by the
-	// above equality checks.
-	if uint64(batch.Batch.EpochNum)+bq.config.SeqWindowSize < batch.L1InclusionBlock.Number {
-		bq.log.Debug("Batch submitted outside sequence window", "epoch", batch.Batch.EpochNum, "inclusion_block", batch.L1InclusionBlock.Number)
-		return false, nil
-	}
-	return true, nil
+	bq.batches[batch.Timestamp] = append(bq.batches[batch.Timestamp], &data)
 }
 
-// deriveBatches pulls a single batch eagerly or a collection of batches if it is the end of
-// the sequencing window.
-func (bq *BatchQueue) deriveBatches(ctx context.Context, l2SafeHead eth.L2BlockRef) ([]*BatchData, error) {
+// deriveNextBatch derives the next batch to apply on top of the current L2 safe head,
+// following the validity rules imposed on consecutive batches,
+// based on currently available buffered batch and L1 origin information.
+// If no batch can be derived yet, then (nil, io.EOF) is returned.
+func (bq *BatchQueue) deriveNextBatch(ctx context.Context) (*BatchData, error) {
 	if len(bq.l1Blocks) == 0 {
-		return nil, io.EOF
+		panic("cannot derive next batch, no origin was prepared")
```

Comment:
Are you sure a `panic` in the hot code path is a good idea? Just worried about a bug triggering this since there is a bunch of slicing of `bq.l1Blocks`. Could it return a scary error instead?

Reply:
Will change it to a scary `NewCriticalError`

Reply:
This one looks to be still not resolved

Reply:
Ah, good catch

Reply:
Confused it with a different panic that needed to be fixed in the same way